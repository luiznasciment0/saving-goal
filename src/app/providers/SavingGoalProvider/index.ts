import constate from 'constate'
import { useReducer, useCallback } from 'react'

import { MinDate, SavingGoalState, Actions } from 'app/models/SavingGoal'

const currencyMask = (money: number) => {
  return new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: 'USD',
    minimumFractionDigits: 2
  }).format(money)
}

const reducer = (state: SavingGoalState, action: Actions) => {
  switch (action.type) {
    case 'SET_TOTAL_AMOUNT':
      return { ...state, totalAmount: action.money }
    case 'SET_DATE':
      return { ...state, reachDate: action.date }
    case 'SET_MONTHLY_AMOUNT':
      return {
        ...state,
        monthlyAmount: action.amount
      }
  }
}

const useSavingGoal = ({ initialDate }: MinDate) => {
  const [state, dispatch] = useReducer(reducer, {
    reachDate: initialDate,
    totalAmount: '',
    monthlyAmount: ''
  })

  const setAmount = (money: string) => {
    dispatch({ type: 'SET_TOTAL_AMOUNT', money })
  }

  const setDateToReachGoal = (date: Date) => {
    dispatch({ type: 'SET_DATE', date })
  }

  const calcMonthlyAmount = useCallback(
    (dateDiff: number, totalAmount: string) => {
      const initialAmount = totalAmount === '0' ? '$0,00' : `${totalAmount}`

      if (dateDiff === 0) {
        dispatch({ type: 'SET_MONTHLY_AMOUNT', amount: initialAmount })
        return
      }

      const numberTotalAmount = Number(totalAmount.replace(/[,]+/g, ''))

      dispatch({
        type: 'SET_MONTHLY_AMOUNT',
        amount: currencyMask(numberTotalAmount / dateDiff)
      })
    },
    []
  )

  return {
    state,
    effects: {
      setAmount,
      setDateToReachGoal,
      calcMonthlyAmount
    }
  }
}

const [SavingGoalProvider, useStateSavingGoal, useEffectsSavingGoal] = constate(
  useSavingGoal,
  (value) => value.state,
  (value) => value.effects
)

export { SavingGoalProvider, useStateSavingGoal, useEffectsSavingGoal }
