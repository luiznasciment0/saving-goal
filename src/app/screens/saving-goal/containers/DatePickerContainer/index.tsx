import { useState } from 'react'

import addMonths from 'date-fns/addMonths'
import subMonths from 'date-fns/subMonths'
import useKeyPress from 'app/hooks/useKeyPress'
import { useEffectsSavingGoal } from 'app/providers/SavingGoalProvider'

import DatePicker from '../../components/DatePicker'

const initialYear = new Date().getFullYear()
const initialMonth = new Date().getMonth() + 1
const initialDate = new Date(initialYear, initialMonth)

const DatePickerContainer = () => {
  const { setDateToReachGoal } = useEffectsSavingGoal()

  const [date, setDate] = useState(initialDate)

  const month = date.toLocaleString('en-US', {
    month: 'long'
  })

  const year = date.getFullYear()

  const incrementDate = () => {
    setDate(addMonths(date, 1))
    setDateToReachGoal(addMonths(date, 1))
  }

  const decrementDate = () => {
    if (date > initialDate) {
      setDate(subMonths(date, 1))
      setDateToReachGoal(subMonths(date, 1))
    }
  }

  useKeyPress('ArrowRight', incrementDate)
  useKeyPress('ArrowLeft', decrementDate)

  return (
    <DatePicker
      incrementDate={incrementDate}
      decrementDate={decrementDate}
      month={month}
      year={year}
    />
  )
}

export default DatePickerContainer
