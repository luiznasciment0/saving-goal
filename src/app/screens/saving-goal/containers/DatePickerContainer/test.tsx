import { render, screen } from '@testing-library/react'

import { SavingGoalProvider } from 'app/providers/SavingGoalProvider'
import DatePickerContainer from '.'

describe('<DatePickerContainer />', () => {
  it('should add a month when righ arrow is pressed', () => {
    const { container } = render(
      <SavingGoalProvider initialDate={new Date(2021, 0)}>
        <DatePickerContainer initialDate={new Date(2021, 0)} />
      </SavingGoalProvider>
    )

    const month = screen.getByRole('heading', { name: /January/i })
    expect(month).toHaveTextContent('January')

    const year = screen.getByRole('heading', { name: /2021/i })
    expect(year).toHaveTextContent('2021')

    expect(container.firstChild).toMatchSnapshot()
  })
})
