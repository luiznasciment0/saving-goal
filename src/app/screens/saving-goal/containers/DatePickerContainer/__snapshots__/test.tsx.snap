// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[
  `<DatePickerContainer /> should add a month when righ arrow is pressed 1`
] = `
.c0 {
  font-size: 1.2rem;
  line-height: 1.6rem;
  font-family: 'Work Sans';
  color: #1e2a32;
  font-weight: 400;
}

<h4
  class="c0"
  color="primaryBlack"
>
  Reach goal by
</h4>
`
